metadata-io/src/main/java/com/linkedin/metadata/changeprocessor/ChangeStreamProcessor.java

```diff
@@ -0,0 +1,106 @@
+package com.linkedin.metadata.changeprocessor;
+
+import com.linkedin.data.template.RecordTemplate;
+import javax.annotation.Nonnull;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+
+/**
+ * Responsible for routing events to the correct registered change processors. Processors are registered based on the
+ * entity/aspect they apply to. If none are supplied then the proposed change is returned.
+ */
+@Slf4j
+public class ChangeStreamProcessor {
+
+  private final Comparator<ChangeProcessor> _processorComparator = Comparator.comparing(ChangeProcessor::getPriority);
+  private final Map<String, SortedSet<ChangeProcessor>> _preProcessors = new HashMap<>();
+  private final Map<String, SortedSet<ChangeProcessor>> _postProcessors = new HashMap<>();
+
+  public void registerPreProcessor(String entityName, String aspectName, ChangeProcessor processor) {
+    registerProcessor(entityName, aspectName, processor, _preProcessors);
+  }
+
+  public void registerPostProcessor(String entityName, String aspectName, ChangeProcessor processor) {
+    registerProcessor(entityName, aspectName, processor, _postProcessors);
+  }
+
+  public ChangeResult preProcessChange(String entityName, String aspectName, RecordTemplate previousAspect,
+      RecordTemplate newAspect) {
+    return process(entityName, aspectName, previousAspect, newAspect, _preProcessors);
+  }
+
+  public ChangeResult postProcessChange(String entityName, String aspectName, RecordTemplate previousAspect,
+      RecordTemplate newAspect) {
+    return process(entityName, aspectName, previousAspect, newAspect, _postProcessors);
+  }
+
+  private void registerProcessor(String entityName, String aspectName, ChangeProcessor processor,
+      Map<String, SortedSet<ChangeProcessor>> processorMap) {
+    String processorKey = (entityName + "/" + aspectName).toLowerCase();
+
+    if (processorMap.containsKey(processorKey)) {
```

Comment:
Does this need to be thread-safe, or is this guaranteed to be called by a single thread only?

Reply:
Good spot, changed to ConcurrentHashMap

Reply:
That does not solve the problem as you are reading (`containsKey`) and writing (`get`) non-atomically, so there can still occur a race condition. You should use something like `computeIfAbsent` to add a `TreeSet` atomically. Then, `TreeSet` itself needs to be thread-safe, so maybe wrap it with `Collections.synchronizedSortedSet()`.

I can sketch something out if you like.

Reply:
Would something like `    processorMap.computeIfAbsent(
        processorKey,
        k -> Collections.synchronizedSortedSet(new TreeSet<>(_processorComparator))
    ).add(processor);` work here?

Reply:
yes, that should be thread-safe